src/azul/deployment.py

```diff
@@ -434,8 +437,46 @@ def client(self, *args, **kwargs):
         Caching the result of this function is not necessary and will be harmful
         if the cached value is used by a thread other than the one that called
         this function.
+
+        :param azul_logging: whether to log the client's requests and responses
+        """
+        client = self.boto3_session.client(*args, **kwargs)
+        if azul_logging:
+            client.meta.events.register('before-send', self._log_client_request)
+            client.meta.events.register('response-received', self._log_client_response)
+        return client
+
+    def _log_client_request(self,
+                            event_name: str,
+                            request: AWSPreparedRequest,
+                            **kwargs: Any
+                            ) -> Optional[AWSResponse]:
+        """
+        Log the request's AWS API action, HTTP method, URL, and body (unless
+        the body is streamable).
+        """
+        event_name = event_name.replace('before-send', 'client')
+        log.info('%s requesting %s to %s', event_name, request.method, request.url)
+        if isinstance(request.body, (str, bytes, bytearray)):
+            log.info('… with request body %r', trunc_ellipses(request.body, 128))
+        return None
+
+    def _log_client_response(self,
+                             *,
+                             event_name: str,
+                             **kwargs: Any
+                             ) -> Optional[AWSResponse]:
+        """
+        Log the request's AWS API action and the response's HTTP status and
+        body (unless the body is streamable).
         """
-        return self.boto3_session.client(*args, **kwargs)
+        event_name = event_name.replace('response-received', 'client')
+        response = kwargs['response_dict']
+        body = response.get('body', '')
+        if isinstance(body, (str, bytes, bytearray)):
+            body = '' if body == b'{}' else f' with {trunc_ellipses(body, 128)!r}'
+        log.info('Got %s from %s%s', response['status_code'], event_name, body)
```

Comment:
Rethink this section. It strikes me as to complicated.  Write down every possible case and what gets logged in that case. The first case is when `response` doesn't have a `body` property. Your implementation conflates that with the case where `response` is `{…, 'body': '', …}`. The comparison to `b'{}'` also smells fishy. I assume that's an empty JSON object. Your implementation doesn't log that, and that's wrong. The equivalent b'{ }' with a space is the serialized representation of an equal JSON structure, but would fail the comparison. Lastly, I don't like that you are taking two different approaches when logging the response body compared to logging the request body. The latter uses a second log statement, the former composes a single log message. Why the difference?

Reply:
Discussed in PL.